crates/trippy-tui/src/locale.rs

```diff
@@ -1,43 +1,125 @@
-/// is not supported, the fallback locale is used.
+/// If the given locale is `None` the system locale is tried. If the system locale cannot be
+/// determined then the fallback locale is used.
 ///
-/// In both cases, the language part of the locale is used if the full locale is not supported.
-pub fn set_locale(locale: Option<&str>) {
-    if let Some(locale) = locale {
-        set_locale_inner(locale);
-    } else if let Some(locale) = sys_locale::get_locale().as_ref() {
-        set_locale_inner(locale);
+/// In all cases, the language part of the locale is used if the full locale is not supported.
+pub fn set_locale(locale: Option<&str>) -> String {
+    let new_locale = if let Some(locale) = locale {
+        String::from(locale)
+    } else if let Some(locale) = sys_locale::get_locale() {
+        locale
     } else {
-        set_locale_inner(FALLBACK_LOCALE);
-    }
+        String::from(FALLBACK_LOCALE)
+    };
+    set_locale_inner(new_locale)
 }
 
-/// Get the current locale.
-pub fn locale() -> String {
-    rust_i18n::locale().to_string()
+/// Get the available locales.
+pub fn available_locales() -> Vec<&'static str> {
+    data()
+        .0
+        .iter()
+        .flat_map(|(_, v)| v.0.keys().map(AsRef::as_ref))
+        .unique()
+        .sorted_unstable()
+        .collect::<Vec<_>>()
 }
 
-/// Get all available locales.
-pub fn available_locales() -> Vec<&'static str> {
-    rust_i18n::available_locales!()
+/// A macro to translate an item to the current locale.
+#[macro_export]
+macro_rules! t {
+    ($key:expr) => {
+        std::borrow::Cow::Borrowed($crate::locale::__translate($key))
+    };
+    ($key:expr, $($kt:ident = $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", stringify!($kt), "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+    ($key:expr, $($kt:literal => $kv:expr),+) => {
+        {
+            let string = t!($key);
+            $(
+                let string = string.replace(concat!("%{", $kt, "}"), &$kv.to_string());
+            )+
+            string
+        }
+    };
+}
+
+/// Translate an item to the current locale.
+///
+/// This function is public as it is used by the `t!` macro, however is not considered part of the
+/// public interface.
+#[doc(hidden)]
+pub fn __translate(item: &str) -> &str {
+    translate_locale(item, locale().load().as_str())
+}
+
+/// Translate an item to a specific locale.
+///
+/// If the item does not exists, the key is returned. Otherwise, if item does not contain the
+/// locale, the fallback locale is used. If the fallback locale does not exist, the key is
+/// returned.
+fn translate_locale<'a>(item: &'a str, locale: &'_ str) -> &'a str {
+    if let Some(key) = data().0.get(item) {
+        if let Some(value) = key.0.get(locale) {
+            value
+        } else if let Some(value) = key.0.get(&split_locale(locale)) {
```

Comment:
reuse common split logic.  aside: we may (later..) want to introduce a "proper" `Locale` type which enforces the rules for locale strings.

Reply:
This can maybe be updated with the `let Some(key) = data().0.get(item) else { return item };` syntax, reducing rightward drift a bit, but otherwise it's really only aesthetic.